
<!DOCTYPE html>


<html lang="en" data-content_root="../../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Predict protein secondary structure &#8212; Chem 193</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  <!--
    this give us a css class that will be invisible only if js is disabled
  -->
  <noscript>
    <style>
      .pst-js-only { display: none !important; }

    </style>
  </noscript>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../../_static/styles/theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />
<link href="../../../_static/styles/pydata-sphinx-theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />

    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=8f2a1f02" />
    <link rel="stylesheet" type="text/css" href="../../../_static/styles/sphinx-book-theme.css?v=a3416100" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../../../_static/nbsphinx-code-cells.css?v=2aa19091" />
  
  <!-- So that users can add custom icons -->
  <script src="../../../_static/scripts/fontawesome.js?digest=8878045cc6db502f8baf"></script>
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf" />
<link rel="preload" as="script" href="../../../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf" />

    <script src="../../../_static/documentation_options.js?v=e645c8fa"></script>
    <script src="../../../_static/doctools.js?v=9bcbadda"></script>
    <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../../../_static/copybutton.js?v=f281be69"></script>
    <script src="../../../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'homework/2-protein-secondary-structure/script/main';</script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="prev" title="Processing protein multiple sequence alignments" href="../../1-python-basics/main.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <meta name="docsearch:version" content="" />
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <dialog id="pst-search-dialog">
    
<form class="bd-search d-flex align-items-center"
      action="../../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
  </dialog>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <dialog id="pst-primary-sidebar-modal"></dialog>
      <div id="pst-primary-sidebar" class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../../../index.html">
  
  
  
  
  
  
    <p class="title logo__title">Chem 193</p>
  
</a></div>
        <div class="sidebar-primary-item">

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../../schedule.html">Schedule</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../syllabus.html">Syllabus</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../lecture.html">Lecture Slides</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../../tutorial/index.html">Tutorials</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../tutorial/python-basics.html">Python Basics</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../tutorial/numpy-jax.html">Numpy &amp; JAX</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../tutorial/linear-regression.html">Linear Regression</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../tutorial/hpc.html">High Performance Computing</a></li>
</ul>
</details></li>
<li class="toctree-l1 current active has-children"><a class="reference internal" href="../../index.html">Homeworks</a><details open="open"><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../../0-linear-algebra.html">Linear Algebra</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../1-python-basics/main.html">Protein MSA</a></li>
<li class="toctree-l2 current active"><a class="current reference internal" href="#">Protein secondary structure</a></li>
</ul>
</details></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
      <div class="sidebar-primary-item">
<div id="ethical-ad-placement"
      class="flat"
      data-ea-publisher="readthedocs"
      data-ea-type="readthedocs-sidebar"
      data-ea-manual="true">
</div></div>
  </div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">


<a href="https://github.com/DingGroup/Chem-193" target="_blank"
   class="btn btn-sm btn-source-repository-button"
   title="Source repository"
   data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fab fa-github"></i>
  </span>

</a>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../../../_sources/homework/2-protein-secondary-structure/script/main.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button>


<button class="btn btn-sm pst-navbar-icon search-button search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
</button>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Predict protein secondary structure</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Model">Model</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#Process-the-training-data">Process the training data</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Define-the-loss-function-for-softmax-regression">Define the loss function for softmax regression</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Train-the-model">Train the model</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Make-predictions-and-compute-the-accuracy-on-the-training-data">Make predictions and compute the accuracy on the training data</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Make-predictions-on-the-test-data">Make predictions on the test data</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Model-interpretation">Model interpretation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Submission-instructions">Submission instructions</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="Predict-protein-secondary-structure">
<h1>Predict protein secondary structure<a class="headerlink" href="#Predict-protein-secondary-structure" title="Link to this heading">#</a></h1>
<p><strong>Author</strong>: YOUR_NAME</p>
<p><strong>Due date</strong>: February 12, 2025, 11:59 PM</p>
<p>In this assignment you will predict the secondary structure of proteins based on their amino acid sequences. The <a class="reference external" href="https://en.wikipedia.org/wiki/Protein_secondary_structure">secondary structure of a protein</a> is the local spatial arrangement of its backbone atoms. It arises from the hydrogen bonds between the backbone atoms and is the building block of proteins’ 3D structure. The most common types of secondary structures are alpha helices, beta sheets, and coils. Given a protein’s 3D structure,
the secondary structure can be determined from the coordinates of its atoms. Depending on the determination method, the secondary structure can be classified into more or fewer types.</p>
<p>Here you will use a dataset of proteins with known secondary structures that were assigned by the <a class="reference external" href="https://pdb-redo.eu/dssp/about">DSSP</a> method based on their 3D structures. Although DSSP assigns eight types of secondary structures, the assignment simplifies them into three types: helix (including types G, H, and I), strand (including types B and E) and coil (including types T, S, ., and P). The training and test datasets are provided in the text files
<a class="reference external" href="../../../_static/protein-secondary-structure/train.txt">train.txt</a> and <a class="reference external" href="../../../_static/protein-secondary-structure/test.txt">test.txt</a>, respectively. Each line starting with a greater-than symbol <code class="docutils literal notranslate"><span class="pre">&gt;</span></code> contains the name of the protein. In the training dataset, the two lines following the name contain the amino acid sequence and the secondary structure, respectively. In the test dataset, only the amino acid sequence is provided. The amino acid sequence is represented by the one-letter
code. The secondary structure is represented as a string of 0 (helix), 1 (strand), and 2 (other).</p>
<p>The task is to train a machine learning model using the training dataset and predict the secondary structure of the proteins using their sequences in the test dataset.</p>
<section id="Model">
<h2>Model<a class="headerlink" href="#Model" title="Link to this heading">#</a></h2>
<p>Different proteins have different numbers of amino acids. If the model were to predict the secondary structure of all amino acids in a protein at once, it would have to handle sequences of different lengths. Although this is possible, this assignment will use a simpler approach. The model will predict the secondary structure of an amino acid based on the amino acids before it and after it. This way, the model will have a fixed-size input and output. Specifically, the model will use as input a
sequence of 15 amino acids and predict the secondary structure of the central amino acid. The model will slide this window along the sequence to predict the secondary structure of all amino acids in a protein. For the amino acids near the beginning and end of the sequence where the window extends beyond the sequence, the model will use a special padding token <code class="docutils literal notranslate"><span class="pre">*</span></code> to complete the window.</p>
<p>As an example, consider the amino acid sequence <code class="docutils literal notranslate"><span class="pre">SIVAGYEVVGSSSASELLSAIEHVAEKA</span></code>. To predict the secondary structure of the starting amino acid <code class="docutils literal notranslate"><span class="pre">S</span></code>, the model will use as input the window <code class="docutils literal notranslate"><span class="pre">*******SIVAGYEV</span></code>. For the second amino acid <code class="docutils literal notranslate"><span class="pre">I</span></code>, the model will use the window <code class="docutils literal notranslate"><span class="pre">******SIVAGYEVV</span></code>. For the residue <code class="docutils literal notranslate"><span class="pre">A</span></code> at position 14, the model will use the window <code class="docutils literal notranslate"><span class="pre">EVVGSSSASELLSAI</span></code>. This process will continue until the model predicts the secondary structure of the last amino acid <code class="docutils literal notranslate"><span class="pre">A</span></code> using the
window <code class="docutils literal notranslate"><span class="pre">IEHVAEKA*******</span></code></p>
<p>Therefore the input to the model will be a sequence of 15 amino acids and the output will be the secondary structure of the central amino acid. The secondary structure of an amino acid is a categorical variable <span class="math notranslate nohighlight">\(y\)</span> with three classes: 0 (helix), 1 (strand), and 2 (other). Therefore, the model will perform a multi-class classification. The input is a sequence of strings, so they need to be converted to numerical values. The simplest way to do this is to use <a class="reference external" href="https://en.wikipedia.org/wiki/One-hot#Machine_learning_and_statistics">one-hot
encoding</a>. In one-hot encoding, each amino acid is represented by a vector of zeros with a one at the position corresponding to the amino acid. For example, if all amino acids (including the padding token <code class="docutils literal notranslate"><span class="pre">*</span></code>) are represented by the list <code class="docutils literal notranslate"><span class="pre">['A',</span> <span class="pre">'C',</span> <span class="pre">'D',</span> <span class="pre">'E',</span> <span class="pre">'F',</span> <span class="pre">'G',</span> <span class="pre">'H',</span> <span class="pre">'I',</span> <span class="pre">'K',</span> <span class="pre">'L',</span> <span class="pre">'M',</span> <span class="pre">'N',</span> <span class="pre">'P',</span> <span class="pre">'Q',</span> <span class="pre">'R',</span> <span class="pre">'S',</span> <span class="pre">'T',</span> <span class="pre">'V',</span> <span class="pre">'W',</span> <span class="pre">'Y',</span> <span class="pre">'*']</span></code>, then the amino acid <code class="docutils literal notranslate"><span class="pre">A</span></code> is represented by the 21-dimensional vector
<code class="docutils literal notranslate"><span class="pre">[1,</span> <span class="pre">0,</span> <span class="pre">...,</span> <span class="pre">0]</span></code>, the amino acid <code class="docutils literal notranslate"><span class="pre">C</span></code> is represented by <code class="docutils literal notranslate"><span class="pre">[0,</span> <span class="pre">1,</span> <span class="pre">0,</span> <span class="pre">...,</span> <span class="pre">0]</span></code>, and so on. The padding token <code class="docutils literal notranslate"><span class="pre">*</span></code> is represented by <code class="docutils literal notranslate"><span class="pre">[0,</span> <span class="pre">0,</span> <span class="pre">...,</span> <span class="pre">1]</span></code>. Because the input is a sequence of 15 amino acids, the input <span class="math notranslate nohighlight">\(x\)</span> to the model is a vector of length <span class="math notranslate nohighlight">\(15 \times 21 = 315\)</span>.</p>
<p>The model will be a softmax regression model, as covered in the lectures.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## Importing the necessary packages</span>
<span class="c1">## If there is any package missing in your uv project, remember to install it using `uv install &lt;package-name&gt;`</span>

<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">jax.numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">jnp</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">jax</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">tqdm</span><span class="w"> </span><span class="kn">import</span> <span class="n">tqdm</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">optax</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">chex</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">optax.tree_utils</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">otu</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">typing</span><span class="w"> </span><span class="kn">import</span> <span class="n">NamedTuple</span>

<span class="n">jax</span><span class="o">.</span><span class="n">config</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="s2">&quot;jax_enable_x64&quot;</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<section id="Process-the-training-data">
<h3>Process the training data<a class="headerlink" href="#Process-the-training-data" title="Link to this heading">#</a></h3>
<p>Given the design of the model, the data in <code class="docutils literal notranslate"><span class="pre">train.txt</span></code> are not directly suitable for training. The data need to be processed to create the input and output pairs for the model. For each sequence in the training data, you need to extract all windows of 15 amino acids and the corresponding secondary structure of the central amino acid. Then you need to convert the amino acids to one-hot encoding. Take the sequence <code class="docutils literal notranslate"><span class="pre">SIVAGYEVVGSSSASELLSAIEHVAEKA</span></code> and its secondary structure
<code class="docutils literal notranslate"><span class="pre">2111111111222000000000220000</span></code> as an example. The first window is <code class="docutils literal notranslate"><span class="pre">*******SIVAGYEV</span></code> and the corresponding secondary structure of the central amino acid <code class="docutils literal notranslate"><span class="pre">S</span></code> is 2. The second window is <code class="docutils literal notranslate"><span class="pre">******SIVAGYEVV</span></code> and the secondary structure of the central amino acid <code class="docutils literal notranslate"><span class="pre">I</span></code> is 1. This process will continue until the last window <code class="docutils literal notranslate"><span class="pre">IEHVAEKA*******</span></code> and the secondary structure of the central amino acid <code class="docutils literal notranslate"><span class="pre">A</span></code> is 0. Each window and the corresponding secondary structure will be an input-output pair for the
model. You need to process all sequences in the training data to create the input-output pairs.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## You need to download the files `train.txt` and `test.txt` mentioned above</span>
<span class="c1">## and place them in the a directory called `data` under the your uv project directory</span>

<span class="n">path_to_train</span> <span class="o">=</span> <span class="s2">&quot;./data/train.txt&quot;</span>


<span class="c1">## read data from train.txt</span>
<span class="n">train_seq</span> <span class="o">=</span> <span class="p">{}</span>

<span class="c1">## finish the code below such that train_seq is a dictionary with the following structure:</span>
<span class="c1">## train_seq[protein_name] = (sequence, secondary_structure),</span>
<span class="c1">## where protein_name is the name of the protein, sequence is the amino acid sequence of the protein, and secondary_structure is the secondary structure of the protein as given in the train.txt file</span>


<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">path_to_train</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
    <span class="c1">############################################</span>
    <span class="c1">#### write your code (5 points) #######</span>
    <span class="c1">############################################</span>
    <span class="kc">None</span>
<br/></pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span><span class="w"> </span><span class="nf">get_windows</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">secondary_structure</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Extract windows of size 15 from the sequence and secondary structure of a protein</span>

<span class="sd">    Args:</span>
<span class="sd">        seq: str, the amino acid sequence of the protein</span>
<span class="sd">        secondary_structure: str, the secondary structure of the protein</span>

<span class="sd">    Returns:</span>
<span class="sd">        list of tuples: each tuple contains (x, y), where x (15x21=315 dimensional) is the one-hot encoding of the amino acid sequence of length 15, and y is the secondary structure of the central amino acid in the window</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1">## amino acid order for one-hot encoding</span>
    <span class="n">amino_acids</span> <span class="o">=</span> <span class="s2">&quot;ACDEFGHIKLMNPQRSTVWY*&quot;</span>

    <span class="c1">############################################</span>
    <span class="c1">#### write your code (15 points) ############</span>
    <span class="c1">############################################</span>

    <span class="k">return</span> <span class="kc">None</span>
</pre></div>
</div>
</div>
<p>Code in the cell below uses the function <code class="docutils literal notranslate"><span class="pre">get_windows</span></code> to process every sequence in the training data and stack both <code class="docutils literal notranslate"><span class="pre">x</span></code> and <code class="docutils literal notranslate"><span class="pre">y</span></code> in numpy arrays: <code class="docutils literal notranslate"><span class="pre">train_xs</span></code> and <code class="docutils literal notranslate"><span class="pre">train_y</span></code>.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#############################################################################</span>
<span class="c1">#### The following code is provided to you and you don&#39;t need to modify it</span>
<span class="c1">#### However, you need to understand it to complete the assignment</span>
<span class="c1">#############################################################################</span>

<span class="c1">## get windows for all proteins in the training set</span>
<span class="c1">## stack all windows in train_xs and train_ys</span>

<span class="n">train_xs</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">train_ys</span> <span class="o">=</span> <span class="p">[]</span>

<span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">ss</span><span class="p">)</span> <span class="ow">in</span> <span class="n">tqdm</span><span class="p">(</span><span class="n">train_seq</span><span class="o">.</span><span class="n">items</span><span class="p">()):</span>
    <span class="n">windows</span> <span class="o">=</span> <span class="n">get_windows</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">ss</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span> <span class="ow">in</span> <span class="n">windows</span><span class="p">:</span>
        <span class="n">train_xs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
        <span class="n">train_ys</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">y</span><span class="p">)</span>

<span class="n">train_xs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">train_xs</span><span class="p">)</span>
<span class="n">train_ys</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">train_ys</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int64</span><span class="p">)</span>

<span class="n">train_xs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">hstack</span><span class="p">((</span><span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">((</span><span class="n">train_xs</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="mi">1</span><span class="p">)),</span> <span class="n">train_xs</span><span class="p">))</span>

<span class="k">assert</span> <span class="n">train_xs</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">==</span> <span class="n">train_ys</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="k">assert</span> <span class="n">train_xs</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="mi">315</span> <span class="o">+</span> <span class="mi">1</span>
<span class="k">assert</span> <span class="n">np</span><span class="o">.</span><span class="n">all</span><span class="p">((</span><span class="n">train_xs</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span> <span class="o">|</span> <span class="p">(</span><span class="n">train_xs</span> <span class="o">==</span> <span class="mi">1</span><span class="p">))</span>
</pre></div>
</div>
</div>
<p>Code in the following cell builds a optimizer, <code class="docutils literal notranslate"><span class="pre">fmin_lbfgs</span></code>, using the <code class="docutils literal notranslate"><span class="pre">optax</span></code> library. You do not need to modify this code. The optimization step will use this optimizer.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span><span class="w"> </span><span class="nf">_run_opt</span><span class="p">(</span><span class="n">init_params</span><span class="p">,</span> <span class="n">fun</span><span class="p">,</span> <span class="n">opt</span><span class="p">,</span> <span class="n">max_iter</span><span class="p">,</span> <span class="n">tol</span><span class="p">):</span>
    <span class="n">value_and_grad_fun</span> <span class="o">=</span> <span class="n">optax</span><span class="o">.</span><span class="n">value_and_grad_from_state</span><span class="p">(</span><span class="n">fun</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">step</span><span class="p">(</span><span class="n">carry</span><span class="p">):</span>
        <span class="n">params</span><span class="p">,</span> <span class="n">state</span> <span class="o">=</span> <span class="n">carry</span>
        <span class="n">value</span><span class="p">,</span> <span class="n">grad</span> <span class="o">=</span> <span class="n">value_and_grad_fun</span><span class="p">(</span><span class="n">params</span><span class="p">,</span> <span class="n">state</span><span class="o">=</span><span class="n">state</span><span class="p">)</span>
        <span class="n">updates</span><span class="p">,</span> <span class="n">state</span> <span class="o">=</span> <span class="n">opt</span><span class="o">.</span><span class="n">update</span><span class="p">(</span>
            <span class="n">grad</span><span class="p">,</span> <span class="n">state</span><span class="p">,</span> <span class="n">params</span><span class="p">,</span> <span class="n">value</span><span class="o">=</span><span class="n">value</span><span class="p">,</span> <span class="n">grad</span><span class="o">=</span><span class="n">grad</span><span class="p">,</span> <span class="n">value_fn</span><span class="o">=</span><span class="n">fun</span>
        <span class="p">)</span>
        <span class="n">params</span> <span class="o">=</span> <span class="n">optax</span><span class="o">.</span><span class="n">apply_updates</span><span class="p">(</span><span class="n">params</span><span class="p">,</span> <span class="n">updates</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">params</span><span class="p">,</span> <span class="n">state</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">continuing_criterion</span><span class="p">(</span><span class="n">carry</span><span class="p">):</span>
        <span class="n">_</span><span class="p">,</span> <span class="n">state</span> <span class="o">=</span> <span class="n">carry</span>
        <span class="n">iter_num</span> <span class="o">=</span> <span class="n">otu</span><span class="o">.</span><span class="n">tree_get</span><span class="p">(</span><span class="n">state</span><span class="p">,</span> <span class="s2">&quot;count&quot;</span><span class="p">)</span>
        <span class="n">grad</span> <span class="o">=</span> <span class="n">otu</span><span class="o">.</span><span class="n">tree_get</span><span class="p">(</span><span class="n">state</span><span class="p">,</span> <span class="s2">&quot;grad&quot;</span><span class="p">)</span>
        <span class="n">err</span> <span class="o">=</span> <span class="n">otu</span><span class="o">.</span><span class="n">tree_l2_norm</span><span class="p">(</span><span class="n">grad</span><span class="p">)</span>
        <span class="k">return</span> <span class="p">(</span><span class="n">iter_num</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span> <span class="o">|</span> <span class="p">((</span><span class="n">iter_num</span> <span class="o">&lt;</span> <span class="n">max_iter</span><span class="p">)</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">err</span> <span class="o">&gt;=</span> <span class="n">tol</span><span class="p">))</span>

    <span class="n">init_carry</span> <span class="o">=</span> <span class="p">(</span><span class="n">init_params</span><span class="p">,</span> <span class="n">opt</span><span class="o">.</span><span class="n">init</span><span class="p">(</span><span class="n">init_params</span><span class="p">))</span>
    <span class="n">final_params</span><span class="p">,</span> <span class="n">final_state</span> <span class="o">=</span> <span class="n">jax</span><span class="o">.</span><span class="n">lax</span><span class="o">.</span><span class="n">while_loop</span><span class="p">(</span>
        <span class="n">continuing_criterion</span><span class="p">,</span> <span class="n">step</span><span class="p">,</span> <span class="n">init_carry</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">final_params</span><span class="p">,</span> <span class="n">final_state</span>


<span class="k">class</span><span class="w"> </span><span class="nc">_InfoState</span><span class="p">(</span><span class="n">NamedTuple</span><span class="p">):</span>
    <span class="n">iter_num</span><span class="p">:</span> <span class="n">chex</span><span class="o">.</span><span class="n">Numeric</span>


<span class="k">def</span><span class="w"> </span><span class="nf">_print_info</span><span class="p">():</span>
    <span class="k">def</span><span class="w"> </span><span class="nf">init_fn</span><span class="p">(</span><span class="n">params</span><span class="p">):</span>
        <span class="k">del</span> <span class="n">params</span>
        <span class="k">return</span> <span class="n">_InfoState</span><span class="p">(</span><span class="n">iter_num</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">update_fn</span><span class="p">(</span><span class="n">updates</span><span class="p">,</span> <span class="n">state</span><span class="p">,</span> <span class="n">params</span><span class="p">,</span> <span class="o">*</span><span class="p">,</span> <span class="n">value</span><span class="p">,</span> <span class="n">grad</span><span class="p">,</span> <span class="o">**</span><span class="n">extra_args</span><span class="p">):</span>
        <span class="k">del</span> <span class="n">params</span><span class="p">,</span> <span class="n">extra_args</span>

        <span class="n">jax</span><span class="o">.</span><span class="n">debug</span><span class="o">.</span><span class="n">print</span><span class="p">(</span>
            <span class="s2">&quot;Iteration: </span><span class="si">{i:&gt;5}</span><span class="s2">, Value: </span><span class="si">{v:&gt;6.3e}</span><span class="s2">, Gradient norm: </span><span class="si">{e:&gt;6.3e}</span><span class="s2">&quot;</span><span class="p">,</span>
            <span class="n">i</span><span class="o">=</span><span class="n">state</span><span class="o">.</span><span class="n">iter_num</span><span class="p">,</span>
            <span class="n">v</span><span class="o">=</span><span class="n">value</span><span class="p">,</span>
            <span class="n">e</span><span class="o">=</span><span class="n">otu</span><span class="o">.</span><span class="n">tree_l2_norm</span><span class="p">(</span><span class="n">grad</span><span class="p">),</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="n">updates</span><span class="p">,</span> <span class="n">_InfoState</span><span class="p">(</span><span class="n">iter_num</span><span class="o">=</span><span class="n">state</span><span class="o">.</span><span class="n">iter_num</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">optax</span><span class="o">.</span><span class="n">GradientTransformationExtraArgs</span><span class="p">(</span><span class="n">init_fn</span><span class="p">,</span> <span class="n">update_fn</span><span class="p">)</span>


<span class="k">def</span><span class="w"> </span><span class="nf">fmin_lbfgs</span><span class="p">(</span><span class="n">initial_params</span><span class="p">,</span> <span class="n">fun</span><span class="p">,</span> <span class="n">max_iter</span><span class="p">,</span> <span class="n">tol</span><span class="p">):</span>
    <span class="n">opt</span> <span class="o">=</span> <span class="n">optax</span><span class="o">.</span><span class="n">chain</span><span class="p">(</span><span class="n">_print_info</span><span class="p">(),</span> <span class="n">optax</span><span class="o">.</span><span class="n">lbfgs</span><span class="p">())</span>
    <span class="k">return</span> <span class="n">_run_opt</span><span class="p">(</span><span class="n">initial_params</span><span class="p">,</span> <span class="n">fun</span><span class="p">,</span> <span class="n">opt</span><span class="p">,</span> <span class="n">max_iter</span><span class="p">,</span> <span class="n">tol</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
</section>
<section id="Define-the-loss-function-for-softmax-regression">
<h2>Define the loss function for softmax regression<a class="headerlink" href="#Define-the-loss-function-for-softmax-regression" title="Link to this heading">#</a></h2>
<p>In the following, you need to complete a few functions to compute the loss for the softmax regression model. The parameter of the loss function <code class="docutils literal notranslate"><span class="pre">loss_fn</span></code> is <code class="docutils literal notranslate"><span class="pre">theta</span></code> which has shape <code class="docutils literal notranslate"><span class="pre">(2,</span> <span class="pre">316)</span></code>, although we have three classes. The first row of <code class="docutils literal notranslate"><span class="pre">theta</span></code> corresponds to the weights for the class 0 and the second row corresponds to the weights for the class 1. Why do we not have a row for class 2? This is because the degenaracy of the softmax function, i.e., the softmax function is invariant to
adding a constant to the scores. Therefore, we can fix one of the scores to zero. In this case, we fix the score of the last class to zero by setting the <code class="docutils literal notranslate"><span class="pre">theta</span></code> parameter for the last class to zero. This is why the shape of <code class="docutils literal notranslate"><span class="pre">theta</span></code> is <code class="docutils literal notranslate"><span class="pre">(2,</span> <span class="pre">316)</span></code> and not <code class="docutils literal notranslate"><span class="pre">(3,</span> <span class="pre">316)</span></code>. Given a <code class="docutils literal notranslate"><span class="pre">theta</span></code> of shape <code class="docutils literal notranslate"><span class="pre">(2,</span> <span class="pre">316)</span></code>, we could always add a row of zeros to make it <code class="docutils literal notranslate"><span class="pre">(3,</span> <span class="pre">316)</span></code>.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span><span class="w"> </span><span class="nf">expand_theta</span><span class="p">(</span><span class="n">theta</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Expand theta by appending a row of zeros at the end</span>

<span class="sd">    Args:</span>
<span class="sd">        theta: jnp.ndarray, the parameters of the model with shape (2, 316)</span>

<span class="sd">    Returns:</span>
<span class="sd">        jnp.ndarray: the expanded theta with shape (3, 316) where the last row is all zeros</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1">############################################</span>
    <span class="c1">#### write your code (5 points) ############</span>

    <span class="k">return</span> <span class="kc">None</span>


<span class="k">def</span><span class="w"> </span><span class="nf">log_likelihood_per_sample</span><span class="p">(</span><span class="n">theta</span><span class="p">,</span> <span class="n">x</span><span class="p">,</span> <span class="n">y</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Compute the log-likelihood of a single sample</span>

<span class="sd">    Args:</span>
<span class="sd">        theta: jnp.ndarray, the parameters of the model with shape (2, 316)</span>
<span class="sd">        x: jnp.ndarray, the input features of one sample</span>
<span class="sd">        y: int, the target label of the sample</span>

<span class="sd">    Returns:</span>
<span class="sd">        float: the log-likelihood of the sample</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1">############################################</span>
    <span class="c1">#### write your code (20 points) ############</span>
    <span class="c1">############################################</span>

    <span class="c1">#### Hint: use the function jax.scipy.special.logsumexp</span>

    <span class="k">return</span> <span class="kc">None</span>


<span class="k">def</span><span class="w"> </span><span class="nf">log_likelihood_data</span><span class="p">(</span><span class="n">theta</span><span class="p">,</span> <span class="n">xs</span><span class="p">,</span> <span class="n">ys</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Compute the log-likelihood of the data by using log_likelihood_per_sample and jax.vmap</span>

<span class="sd">    Args:</span>
<span class="sd">        theta: jnp.ndarray, the parameters of the model with shape (2, 316)</span>
<span class="sd">        xs: jnp.ndarray, the input features of all samples with shape (n, 316)</span>
<span class="sd">        ys: jnp.ndarray, the target labels of all samples with shape (n,)</span>

<span class="sd">    Returns:</span>
<span class="sd">        jnp.ndarray: the log-likelihood of all samples with shape (n,)</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1">############################################</span>
    <span class="c1">#### write your code (5 points) ############</span>
    <span class="c1">############################################</span>

    <span class="k">return</span> <span class="kc">None</span>


<span class="k">def</span><span class="w"> </span><span class="nf">loss_fn</span><span class="p">(</span><span class="n">theta</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;the loss function is defined as the mean negative log-likelihood of the samples,</span>

<span class="sd">    Args:</span>
<span class="sd">        theta: jnp.ndarray, the parameters of the model with shape (2, 316)</span>

<span class="sd">    Returns:</span>
<span class="sd">        float: the mean negative log-likelihood of the samples</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1">############################################</span>
    <span class="c1">#### write your code (5 points) ############</span>
    <span class="c1">############################################</span>

    <span class="k">return</span>
</pre></div>
</div>
</div>
</section>
<section id="Train-the-model">
<h2>Train the model<a class="headerlink" href="#Train-the-model" title="Link to this heading">#</a></h2>
<p>After the data are processed properly and the loss function is defined, it is time to train the model. Thanks to the automatic differentiation provided by JAX, you do not need to write the function to compute the gradient of the loss function. You only need to provide the loss function to the optimizer along with the initial parameters and stopping criteria.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## initialize theta with zeros</span>
<span class="n">m</span> <span class="o">=</span> <span class="mi">3</span>
<span class="n">theta_init</span> <span class="o">=</span> <span class="n">jnp</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">m</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span> <span class="n">train_xs</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]))</span>

<span class="c1">## check the initial value of the loss function and the norm of the gradient at the initial point</span>
<span class="nb">print</span><span class="p">(</span>
    <span class="sa">f</span><span class="s2">&quot;Initial value: </span><span class="si">{</span><span class="n">loss_fn</span><span class="p">(</span><span class="n">theta_init</span><span class="p">)</span><span class="si">:</span><span class="s2">.2e</span><span class="si">}</span><span class="s2"> &quot;</span>
    <span class="sa">f</span><span class="s2">&quot;Initial gradient norm: </span><span class="si">{</span><span class="n">otu</span><span class="o">.</span><span class="n">tree_l2_norm</span><span class="p">(</span><span class="n">jax</span><span class="o">.</span><span class="n">grad</span><span class="p">(</span><span class="n">loss_fn</span><span class="p">)(</span><span class="n">theta_init</span><span class="p">))</span><span class="si">:</span><span class="s2">.2e</span><span class="si">}</span><span class="s2">&quot;</span>
<span class="p">)</span>

<span class="c1">## optimize the loss function using L-BFGS</span>
<span class="n">final_theta</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">fmin_lbfgs</span><span class="p">(</span><span class="n">theta_init</span><span class="p">,</span> <span class="n">loss_fn</span><span class="p">,</span> <span class="n">max_iter</span><span class="o">=</span><span class="mi">300</span><span class="p">,</span> <span class="n">tol</span><span class="o">=</span><span class="mf">1e-4</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span>
    <span class="sa">f</span><span class="s2">&quot;Final value: </span><span class="si">{</span><span class="n">loss_fn</span><span class="p">(</span><span class="n">final_theta</span><span class="p">)</span><span class="si">:</span><span class="s2">.2e</span><span class="si">}</span><span class="s2">, &quot;</span>
    <span class="sa">f</span><span class="s2">&quot;Final gradient norm: </span><span class="si">{</span><span class="n">otu</span><span class="o">.</span><span class="n">tree_l2_norm</span><span class="p">(</span><span class="n">jax</span><span class="o">.</span><span class="n">grad</span><span class="p">(</span><span class="n">loss_fn</span><span class="p">)(</span><span class="n">final_theta</span><span class="p">))</span><span class="si">:</span><span class="s2">.2e</span><span class="si">}</span><span class="s2">&quot;</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="Make-predictions-and-compute-the-accuracy-on-the-training-data">
<h2>Make predictions and compute the accuracy on the training data<a class="headerlink" href="#Make-predictions-and-compute-the-accuracy-on-the-training-data" title="Link to this heading">#</a></h2>
<p>After the model is trained, you need to make predictions on the training data and compute the accuracy of the predictions. For each residue in the training data, you need to compute the probability of each class and assign the class with the highest probability as the predicted class. Then you need to compare the predicted classes with the true classes and compute the accuracy. The accuracy is the number of correct predictions divided by the total number of predictions.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">def</span><span class="w"> </span><span class="nf">predict_per_sample</span><span class="p">(</span><span class="n">theta</span><span class="p">,</span> <span class="n">x</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Predict the class of a single sample</span>

<span class="sd">    Args:</span>
<span class="sd">        theta: jnp.ndarray, the parameters of the model with shape (2, 316)</span>
<span class="sd">        x: jnp.ndarray, the input features of one sample</span>

<span class="sd">    Returns:</span>
<span class="sd">        int: the predicted class of the sample</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1">############################################</span>
    <span class="c1">#### write your code (5 points) ############</span>
    <span class="c1">############################################</span>

    <span class="k">return</span> <span class="kc">None</span>


<span class="k">def</span><span class="w"> </span><span class="nf">predict</span><span class="p">(</span><span class="n">theta</span><span class="p">,</span> <span class="n">xs</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Predict the classes of multiple samples by using predict_per_sample and jax.vmap</span>

<span class="sd">    Args:</span>
<span class="sd">        theta: jnp.ndarray, the parameters of the model with shape (2, 316)</span>
<span class="sd">        xs: jnp.ndarray, the input features of multiple samples with shape (n, 316)</span>

<span class="sd">    Returns:</span>
<span class="sd">        jnp.ndarray: the predicted classes of the samples with shape (n,)</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1">############################################</span>
    <span class="c1">#### write your code (5 points) ############</span>
    <span class="c1">############################################</span>

    <span class="k">return</span> <span class="kc">None</span>


<span class="k">def</span><span class="w"> </span><span class="nf">compute_accuray</span><span class="p">(</span><span class="n">y_pred</span><span class="p">,</span> <span class="n">y_true</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Compute the accuracy of the model</span>

<span class="sd">    Args:</span>
<span class="sd">        y_pred: jnp.ndarray, the predicted classes of the samples with shape (n,)</span>
<span class="sd">        y_true: jnp.ndarray, the true classes of the samples with shape (n,)</span>

<span class="sd">    Returns:</span>
<span class="sd">        float: the accuracy of the model</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1">############################################</span>
    <span class="c1">#### write your code (5 points) ############</span>
    <span class="c1">############################################</span>

    <span class="k">return</span> <span class="kc">None</span>


<span class="n">y_pred</span> <span class="o">=</span> <span class="n">predict</span><span class="p">(</span><span class="n">final_theta</span><span class="p">,</span> <span class="n">train_xs</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Compute the accuracy of the model on the training set using the initial theta and the final theta. The accuracy with the initial theta should be around 33% because the initial theta is zero and the model is making random predictions. The accuracy with the final theta should be significantly higher because the model has learned the weights that minimize the loss function on the training data.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">y_pred_init</span> <span class="o">=</span> <span class="n">predict</span><span class="p">(</span><span class="n">theta_init</span><span class="p">,</span> <span class="n">train_xs</span><span class="p">)</span>
<span class="n">accuracy_init</span> <span class="o">=</span> <span class="n">compute_accuray</span><span class="p">(</span><span class="n">y_pred_init</span><span class="p">,</span> <span class="n">train_ys</span><span class="p">)</span>

<span class="n">y_pred_final</span> <span class="o">=</span> <span class="n">predict</span><span class="p">(</span><span class="n">final_theta</span><span class="p">,</span> <span class="n">train_xs</span><span class="p">)</span>
<span class="n">accuracy_final</span> <span class="o">=</span> <span class="n">compute_accuray</span><span class="p">(</span><span class="n">y_pred_final</span><span class="p">,</span> <span class="n">train_ys</span><span class="p">)</span>

<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Accuracy of the model with initial theta: </span><span class="si">{</span><span class="n">accuracy_init</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Accuracy of the model with final theta: </span><span class="si">{</span><span class="n">accuracy_final</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<br/></pre></div>
</div>
</div>
</section>
<section id="Make-predictions-on-the-test-data">
<h2>Make predictions on the test data<a class="headerlink" href="#Make-predictions-on-the-test-data" title="Link to this heading">#</a></h2>
<p>After the model is trained, you need to make predictions on the test sequences. The test sequences are provided in the file <code class="docutils literal notranslate"><span class="pre">test.txt</span></code>. You need to use the trained model to predict the secondary structure of all residues in the test sequences. The predictions should be written to a text file named <code class="docutils literal notranslate"><span class="pre">test_prediction.txt</span></code> and its format should be the same as the file <code class="docutils literal notranslate"><span class="pre">train.txt</span></code>.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">###########################################################################</span>
<span class="c1">#### write your code here for the task described above (10 points) ########</span>
<span class="c1">###########################################################################</span>
<br/><br/></pre></div>
</div>
</div>
</section>
<section id="Model-interpretation">
<h2>Model interpretation<a class="headerlink" href="#Model-interpretation" title="Link to this heading">#</a></h2>
<p>You have trained the soft-max regression model to predict the secondary structure of proteins. Specifically the model has learned the values of the parameter <code class="docutils literal notranslate"><span class="pre">theta</span></code>.</p>
<p><strong>Question 1.</strong></p>
<p>If someone asks you, “Based on your model, what is the order of preference of the amino acids for the helix secondary structure?” How would you answer this question based on the values of the parameter <code class="docutils literal notranslate"><span class="pre">theta</span></code>?</p>
<p><strong>Type your answer in the cell below.</strong></p>
<p><em>Your answer to Question 1</em>:</p>
<p><strong>Question 2.</strong></p>
<p>Even though the model has been learned to predict the secondary structure of proteins based on the amino acid sequence, is it possible to use the model to design protein sequences so that they have a specific secondary structure? If yes, how would you do it? If no, why not?</p>
<p><strong>Type your answer in the cell below.</strong></p>
<p><em>Your answer to Question 2</em>:</p>
</section>
<section id="Submission-instructions">
<h2>Submission instructions<a class="headerlink" href="#Submission-instructions" title="Link to this heading">#</a></h2>
<ol class="arabic simple">
<li><p>You need to create a folder named <code class="docutils literal notranslate"><span class="pre">assignment-2-protein-secondary-structure</span></code> under the OneDrive folder that has been shared with you.</p></li>
<li><p>Complete the code and answer the questions as described above in this Jupyter notebook. Make sure to save your work and name your Jupyter notebook as <code class="docutils literal notranslate"><span class="pre">main.ipynb</span></code>.</p></li>
<li><p>Upload the <code class="docutils literal notranslate"><span class="pre">main.ipynb</span></code>, <code class="docutils literal notranslate"><span class="pre">pyproject.toml</span></code>, and <code class="docutils literal notranslate"><span class="pre">test_prediction.txt</span></code> files to the <code class="docutils literal notranslate"><span class="pre">assignment-2-protein-secondary-structure</span></code> folder that you created in step 1.</p></li>
</ol>
<p><strong>Note:</strong> Please make sure the names of the folder and files are exactly as instructed.</p>
</section>
</section>


                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="../../1-python-basics/main.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Processing protein multiple sequence alignments</p>
      </div>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <dialog id="pst-secondary-sidebar-modal"></dialog>
                <div id="pst-secondary-sidebar" class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Model">Model</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#Process-the-training-data">Process the training data</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Define-the-loss-function-for-softmax-regression">Define the loss function for softmax regression</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Train-the-model">Train the model</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Make-predictions-and-compute-the-accuracy-on-the-training-data">Make predictions and compute the accuracy on the training data</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Make-predictions-on-the-test-data">Make predictions on the test data</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Model-interpretation">Model interpretation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Submission-instructions">Submission instructions</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Xinqiang Ding
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2024, Xinqiang Ding.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script defer src="../../../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf"></script>
<script defer src="../../../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>